### PR TITLE
Client: Support endpoints that return layout components

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes: #4836

If an endpoint returns a layout component, it will not be returned by the api but all other IOComponents will (excluding State). These components are excluded from the view_api page as well. But the endpoint is still considered valid for the client.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests & Changelog

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

1. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`
  
1. Unless the pull request is labeled with the "no-changelog-update" label by a maintainer of the repo, all pull requests must update the changelog located in `CHANGELOG.md`:

Please add a brief summary of the change to the Upcoming Release section of the `CHANGELOG.md` file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

### CODE REVIEW

**Comment by freddyaboulton** [comment]
UI and gradio unit tests should pass when new version of the client is released

**Comment by abidlabs** [comment]
Instead of making "row", etc. all serializable components, wouldn't a simpler approach be to modify `skip_components` here:

https://github.com/gradio-app/gradio/blob/5dc445b08fd76cacec1354b05a2d44f23d127a30/gradio/blocks.py#L505

to include "row", etc. Then, we could just remove this logic altogether:

https://github.com/gradio-app/gradio/blob/5dc445b08fd76cacec1354b05a2d44f23d127a30/gradio/blocks.py#L520

I think that's more appropriate since the layouts are not actually serializable and we shouldn't treat them as such.

File: client/python/gradio_client/client.py [hunk 1]
```diff
@@ -427,7 +427,7 @@ def view_api(
 
         # Versions of Gradio older than 3.29.0 returned format of the API info
         # from the /info endpoint
-        if version.parse(self.config.get("version", "2.0")) > version.Version("3.29.0"):
+        if version.parse(self.config.get("version", "2.0")) > version.Version("3.36.1"):
```
Comment (akx) [inline client/python/gradio_client/client.py:430]: Is this `>` on purpose (c.f. `>=`)?

File: client/python/gradio_client/utils.py [hunk 2]
```diff
@@ -31,10 +31,22 @@
 CONFIG_URL = "config"
 API_INFO_URL = "info"
 RAW_API_INFO_URL = "info?serialize=False"
-SPACE_FETCHER_URL = "https://gradio-space-api-fetcher-v2.hf.space/api"
+SPACE_FETCHER_URL = "https://freddyaboulton-space-api-fetcher.hf.space/api"
 RESET_URL = "reset"
 SPACE_URL = "https://hf.space/{}"
 
+SKIP_COMPONENTS = [
```
Comment (akx) [inline client/python/gradio_client/utils.py:38]: This should probably be a set for faster lookup.

**Comment by freddyaboulton** [comment]
@abidlabs The skip_components in `blocks.py` are already updated to include state and all the layout components. The reason I included them in the serializer map is to preserve the principle that each output component has a serializer. That way we treat the layout components the same way we do state. I think that it's conceptually sound (they are both returned by the server but we don't want to display to users) and it makes the footprint of this PR pretty small. Although the layout components are in the serializer map, they will not be displayed in the api info or returned to users and the serializing logic is a no-op.

**Comment by abidlabs** [comment]
> The skip_components in blocks.py are already updated to include state and all the layout components

Nice, I missed this. 

But if we are already skipping the layout components, why do we need to add serializers for them? Imo that doesn't make sense because the serializers should only exist for components that can hold values since they define how this value is serialized between different formats. `gr.State` can hold a value, but the layout components don't hold any values. 

Happy to sync if that's easier

**Comment by freddyaboulton** [comment]
> We should instead "skip" the serialization of these components. (Skipping could just mean the identity function)

In `COMPONENT_MAPPING`, I map the layout components to the `SimpleSerializable`. That's the same as the identify function no?

I guess I I could remove the layout components from the `COMPONENT_MAPPING` and instead modify the following code to not trigger the `assert` for the layout components and manually add a `SimpleSerializer` for the layout components. That's equivalent to what the current code does though and feels harder to maintain.

```python
                        assert (
                            component_name in serializing.COMPONENT_MAPPING
                        ), f"Unknown component: {component_name}, you may need to update your gradio_client version."
                        serializer = serializing.COMPONENT_MAPPING[component_name]
                    serializers.append(serializer())  # type: ignore
```

**Comment by abidlabs** [comment]
I think we should replace:

```py
                    else:
                        assert (
                            component_name in serializing.COMPONENT_MAPPING
                        ), f"Unknown component: {component_name}, you may need to update your gradio_client version."
                        deserializer = serializing.COMPONENT_MAPPING[component_name]

```

with
```py
                    elif component_name in utils.SKIP_COMPONENTS:
                         deserializer = serializing.SimpleSerializable
                    else:
                        assert (
                            component_name in serializing.COMPONENT_MAPPING
                        ), f"Unknown component: {component_name}, you may need to update your gradio_client version."
                        deserializer = serializing.COMPONENT_MAPPING[component_name]
```

And then just remove these from serializing.COMPONENT_MAPPING:

```py
    "row": SimpleSerializable,
    "column": SimpleSerializable,
    "tabs": SimpleSerializable,
    "tab": SimpleSerializable,
    "box": SimpleSerializable,
    "form": SimpleSerializable,
    "accordion": SimpleSerializable,
    "group": SimpleSerializable,
```

That reduces the footprint of the PR and should make it even easier to maintain since we only need to add components in one place: SKIP_COMPONENTS.

**Comment by freddyaboulton** [comment]
Followed your suggestion @abidlabs !

File: client/python/gradio_client/utils.py [hunk 3]
```diff
@@ -31,10 +31,22 @@
 CONFIG_URL = "config"
 API_INFO_URL = "info"
 RAW_API_INFO_URL = "info?serialize=False"
-SPACE_FETCHER_URL = "https://gradio-space-api-fetcher-v2.hf.space/api"
+SPACE_FETCHER_URL = "https://freddyaboulton-space-api-fetcher.hf.space/api"
```
Comment (abidlabs) [inline client/python/gradio_client/utils.py:34]: We just need to remember to update this

File: client/python/test/test_serializing.py [hunk 4]
```diff
@@ -22,7 +22,9 @@ def test_duplicate(serializer_class):
 def test_check_component_fallback_serializers():
     for component_name, class_type in COMPONENT_MAPPING.items():
         # skip components that cannot be instantiated without parameters
-        if component_name in ["dataset", "interpretation"]:
+        if component_name in ["dataset", "interpretation"] + list(
+            SKIP_COMPONENTS - {"state"}
```
Comment (abidlabs) [inline client/python/test/test_serializing.py:26]: Should we just add `["dataset", "interpretation"]` to SKIP_COMPONENTS and remove them from COMPONENT_MAPPING?

**Comment by abidlabs** [comment]
Small nit above. Code looks good + tested. LGTM @freddyaboulton

**Review by abidlabs — APPROVED** [review]

**Comment by freddyaboulton** [comment]
Thanks for the review @abidlabs !!

**Pull request merged**